pkg/reader/usdc_reader.go

```diff
@@ -0,0 +1,307 @@
+	// We need entire MessageSent payload to use that with the Attestation API
+	cr, ok := u.boundContracts[source]
+	if !ok {
+		return nil, fmt.Errorf("no contract bound for chain %d", source)
+	}
+
+	iter, err := u.contractReaders[source].QueryKey(
+		ctx,
+		cr,
+		query.KeyFilter{
+			Key:         consts.EventNameCCTPMessageSent,
+			Expressions: []query.Expression{},
+		},
+		query.NewLimitAndSort(query.Limit{}, query.NewSortBySequence(query.Asc)),
+		&MessageSentEvent{},
+	)
+	if err != nil {
+		return nil, fmt.Errorf("error querying contract reader for chain %d: %w", source, err)
+	}
+
+	messageSentEvents := make(map[eventID]MessageHash)
+	for _, item := range iter {
+		event, ok1 := item.Data.(*MessageSentEvent)
+		if !ok1 {
+			return nil, fmt.Errorf("failed to cast %v to Message", item.Data)
+		}
+		e, err1 := event.unpackID()
+		if err1 != nil {
+			return nil, err1
+		}
+		messageSentEvents[e] = event.Arg0
+	}
+
+	// 3. This should be done by ChainReader - picking only events matching eventIDs.
+	// Right now ChainReader doesn't support filtering by specific data words
+	out := make(map[exectypes.MessageTokenID]MessageHash)
+	for tokenID, messageID := range eventIDs {
+		messageHash, ok1 := messageSentEvents[messageID]
+		if !ok1 {
+			// Token not available in the source chain
+			continue
+		}
+		out[tokenID] = messageHash
+	}
+
+	return out, nil
+}
+
+func (u usdcMessageReader) recreateMessageTransmitterEvents(
+	destChainSelector cciptypes.ChainSelector,
+	tokens map[exectypes.MessageTokenID]cciptypes.RampTokenAmount,
+) (map[exectypes.MessageTokenID]eventID, error) {
+	messageTransmitterEvents := make(map[exectypes.MessageTokenID]eventID)
+
+	for id, token := range tokens {
+		sourceTokenPayload, err := NewSourceTokenDataPayloadFromBytes(token.ExtraData)
+		if err != nil {
+			return nil, err
+		}
+
+		// USDC message payload:
+		// uint32 _msgVersion,
+		// uint32 _msgSourceDomain,
+		// uint32 _msgDestinationDomain,
+		// uint64 _msgNonce,
+		// bytes32 _msgSender,
+		// Since it's packed, all of these values contribute to the first slot
+
+		msgVersionBytes := [4]byte{}
+		binary.BigEndian.PutUint32(msgVersionBytes[:], CCTPMessageVersion)
+
+		sourceDomainBytes := [4]byte{}
+		binary.BigEndian.PutUint32(sourceDomainBytes[:], sourceTokenPayload.SourceDomain)
+
+		destDomain, ok := u.cctpDestDomain[uint64(destChainSelector)]
+		if !ok {
+			return nil, fmt.Errorf("destination domain not found for chain %s", destChainSelector)
+		}
+
+		destDomainBytes := [4]byte{}
+		binary.BigEndian.PutUint32(destDomainBytes[:], destDomain)
+
+		nonceBytes := [8]byte{}
+		binary.BigEndian.PutUint64(nonceBytes[:], sourceTokenPayload.Nonce)
+
+		senderBytes := [12]byte{}
+
+		var buf bytes.Buffer
+		buf.Write(msgVersionBytes[:])
+		buf.Write(sourceDomainBytes[:])
+		buf.Write(destDomainBytes[:])
+		buf.Write(nonceBytes[:])
+		buf.Write(senderBytes[:])
+		messageTransmitterEvents[id] = [32]byte(buf.Bytes()[:32])
```

Comment:
Found another option here, you can combine the appending with the BigEndian call
```suggestion
		destDomain, ok := u.cctpDestDomain[uint64(destChainSelector)]
		if !ok {
			return nil, fmt.Errorf("destination domain not found for chain %s", destChainSelector)
		}
		
		var buf []byte
		buf = binary.BigEndian.AppendUint32(buf, CCTPMessageVersion)
		buf = binary.BigEndian.AppendUint32(buf, sourceTokenPayload.SourceDomain)
		buf = binary.BigEndian.AppendUint32(buf, destDomain)
		buf = binary.BigEndian.AppendUint64(buf, sourceTokenPayload.Nonce)
		buf = append(buf2, v_senderBytes[:]...)
		
		messageTransmitterEvents[id] = [32]byte(buf[:32])
```